python_modules/libraries/dagster-components/dagster_components/resolved/base.py

```diff
@@ -0,0 +1,271 @@
+        of the derived model loaded from schema compliant yaml and
+        evaluating any template strings.
+
+    The fields/__init__ arguments of the class can be Annotated with
+    Resolver to customize the resolution or model derivation.
+    """
+
+    @classmethod
+    def model(cls) -> type[ResolvableModel]:
+        return derive_model_type(cls)
+
+    @classmethod
+    def resolve_from_model(cls, context: "ResolutionContext", model: ResolvableModel):
+        return cls(**_resolve_fields(model, cls, context))
+
+    @classmethod
+    def resolve_from_yaml(cls, yaml: str):
+        parsed_and_src_tree = parse_yaml_with_source_positions(yaml)
+        model = _parse_and_populate_model_with_annotated_errors(
+            cls=cls.model(),
+            obj_parse_root=parsed_and_src_tree,
+            obj_key_path_prefix=[],
+        )
+        # support adding scopes?
+        context = ResolutionContext.default(parsed_and_src_tree.source_position_tree)
+        return cls.resolve_from_model(context, model)
+
+
+class _Unset: ...
+
+
+def derive_model_type(
+    target_type: type[Resolvable],
+) -> type[ResolvableModel]:
+    if target_type not in _DERIVED_MODEL_REGISTRY:
+        model_name = f"{target_type.__name__}Model"
+
+        model_fields: dict[
+            str, Any
+        ] = {}  # use Any to appease type checker when **-ing in to create_model
+
+        for name, (annotation, has_default) in _get_annotations(target_type).items():
+            field_resolver = _get_resolver(annotation, name)
+            field_name = field_resolver.model_field_name or name
+            field_type = field_resolver.model_field_type or annotation
+
+            field_infos = []
+            if has_default:
+                field_infos.append(Field(default=_Unset))
+
+            model_fields[field_name] = (
+                field_type,
+                FieldInfo.merge_field_infos(*field_infos),
+            )
+
+        try:
+            _DERIVED_MODEL_REGISTRY[target_type] = create_model(
+                model_name,
+                __base__=ResolvableModel,
+                **model_fields,
+            )
+        except PydanticSchemaGenerationError as e:
+            raise ResolutionException(f"Unable to derive Model for {target_type}") from e
+
+    return _DERIVED_MODEL_REGISTRY[target_type]
+
+
+def _get_annotations(resolved_type: type[Resolvable]):
+    annotations = {}
+    if is_dataclass(resolved_type):
+        for f in fields(resolved_type):
+            has_default = f.default is not MISSING or f.default_factory is not MISSING
+            annotations[f.name] = (f.type, has_default)
+        return annotations
+    elif _safe_is_subclass(resolved_type, BaseModel):
+        for name, field_info in resolved_type.model_fields.items():
+            annotations[name] = (field_info.rebuild_annotation(), field_info.is_required)
+        return annotations
+    # can update to support:
+    # * classes w/ __init__
```

Comment:
I think it is important to handle arbitrary annotations.  Here's AI-written [code](https://grok.com/share/bGVnYWN5_f7d0baa7-e729-4af0-b437-bd0975b14276) to do it.

Reply:
adding it in the next pr upstack